### ISSUE TITLE
Consider forming A^-1 explicitly in log likelihood computations

### ISSUE THREAD

**Issue by suntzu86** [Yelp/MOE#180]
The gradients and hessians of Log Marginal Likelihood & Leave One Out Log Likelihood (and likely other log likelihood measures) can be computed much faster. For leave one out, using explicit inverse is way faster even for just the log likelihood computation.
- For gradient descent, the dominant term is about `O(N_{sampled}^3)` for explicit inverse and `O(N_{hyper}*N_{sampled}^3)` for the backsolve version.
- For newton, the dominant term is about `O(N_{sampled}^3)` for explicit inverse and `O(N_{hyper}^2 * N_{sampled}^3)` for backsolve.

Practically, hyperparam opt runs 2-3x+ faster even on relatively small problems (num_sampled = 100, dim = 2) with gradient descent or newton.

However, for ill-conditioned matrices, the loss of accuracy can be quite large. So it would be advantageous to perform some iterations with the faster (but numerically unstable) explicit A^-1 computation, and then use the backsolve methods to provide corrections. Two ideas:
1. Out of N optimization steps, do the first fraction with explicit inverses and then the remainder with backsolve
2. Run m backsolve computation steps for every n explicit inverse steps. Could be 1:1, 2:3, or whatever.

Or there may even be a (cheap!) way to detect when loss of precision is likely and only switch to backsolves when it is needed.

For example, if the noise variance is ~ 0.1, the condition number of the covariance matrix (recall it's SPD) cannot be more than ~100. In such a case, the explicit inverse would likely behave reasonably.
